### PR TITLE
Handle missing branches in `BinOp`

### PR DESCRIPTION
**Pull request by vthemelis**
Fixes https://github.com/agronholm/typeguard/issues/384 .

The issue is due to the `left` or `right` attributes of `BinOp` in the case that the `AnnotationTransformer` evaluates any of them to be `None`.

The code that removes the attributes completely is:
https://github.com/python/cpython/blob/e407cea1938b80b1d469f148a4ea65587820e3eb/Lib/ast.py#L493-L498

This patch removes the crush but I think shows another bug where in the test I added, the argument `A` is expected to only have a value of `None`. NB that this is not just wrong but also inconsistent with what happens when using `Union` instead of `|`

### CODE REVIEW

File: tests/test_transformer.py [hunk 1]
```diff
@@ -1570,6 +1570,68 @@ def foo(x: str) -> None:
...
+        unparse(node)
+        == dedent(
+            """
+            from __future__ import annotations
+
+            def foo(A: A | None) -> None:
+                from typeguard import TypeCheckMemo
+                from typeguard._functions import check_argument_types
+                memo = TypeCheckMemo(globals(), locals())
+                check_argument_types('foo', {'A': (A, None)}, memo)
```
Comment (vthemelis) [inline tests/test_transformer.py:1629]: This doesn't look right and it's inconsistent with what we get when using `Union` like in the previous test.

Perhaps I should change the code to return `None` if either of the `BinOp` branches has been erased for now?

Comment (agronholm) [reply tests/test_transformer.py:1629]: I'll take a good look at this tomorrow. Some notes:

- The first test passes without your patch
- Did you intentionally name the class and the argument the same? I'm not sure how that makes any sense.
- What, in your opinion, should the result look like?

Comment (vthemelis) [reply tests/test_transformer.py:1629]: Hi, thanks for the quick review!
> The first test passes without your patch
Yes, I added the first test to compare the output of `Union` and `|` to see if they are consistent. It can be removed but I think that it helps with the conversation.

> Did you intentionally name the class and the argument the same? I'm not sure how that makes any sense.
Yes, I did that intentionally as the easiest way to get a `None` transformation of either branch of the `BinOp`.

> What, in your opinion, should the result look like?
It should look the same as the one for `Union` I think. I do understand that there are some issues with shadowed types though: https://github.com/agronholm/typeguard/issues/337

Comment (vthemelis) [reply tests/test_transformer.py:1629]: Btw, I haven't tried it yet but maybe one way to get access to the shadowed type inside the function is to do something like:
```
from typing import Type

Type[A].__origin__
```

Comment (vthemelis) [reply tests/test_transformer.py:1629]: Interesting:
```Python
@typechecked
def foo(int: Union[int, bool]) -> int:
    return int

foo(42)
```
gives
```
TypeError: Union[arg, ...]: each arg must be a type. Got 42.
```

So I guess that the behaviour with `Union` is also incorrect due to the shadowing of the type.

Comment (agronholm) [reply tests/test_transformer.py:1629]: How about this: cut out the first test, leave the second in. This PR does demonstrably fix a bug, so let's just get this in. If you'd just add a note in the changelog, I'll merge. Any other union related behaviors should be addressed in a separate issue/PR.

Comment (vthemelis) [reply tests/test_transformer.py:1629]: I don't think that this PR 'fixes' a bug as it is; it only replaces one bug for another one.

In the example above, only `None` is an accepted value for `foo` whereas probably only value should be accepted (as we should effectively handle `int` as `Any` in this case.

This is demonstrated on this test:

https://github.com/agronholm/typeguard/pull/392/commits/5b5f91f3c5926fd2697b8381813c187e3c526548#diff-00ca1530aa0e4a5f0915df54f28fd180973f39b478af9b8beca7b49884903cd5R675-R685

(currently failing)

Comment (agronholm) [reply tests/test_transformer.py:1629]: I think that needs a separate bugfix. This PR could be merged to just fix the transformer issue.

Comment (vthemelis) [reply tests/test_transformer.py:1598]: @agronholm, I removed the failing test and the classic union test. I'll add an issue about the `|` bug after we merge this.

File: docs/versionhistory.rst [hunk 2]
```diff
@@ -3,6 +3,11 @@ Version history
 
 This library adheres to `Semantic Versioning 2.0 <https://semver.org/#semantic-versioning-200>`_.
 
+**unreleased**
```
Comment (agronholm) [inline docs/versionhistory.rst:6]:
```suggestion
**UNRELEASED**
```

File: docs/versionhistory.rst [hunk 3]
```diff
@@ -3,6 +3,11 @@ Version history
...
+
+- Fixed handling of `BinOp` nodes with missing branches in AnnotationTransformer
```
Comment (agronholm) [inline docs/versionhistory.rst:8]:
```suggestion
- Fixed ``AttributeError`` where the transformer removed elements from a PEP 604 union
```

File: src/typeguard/_transformer.py [hunk 4]
```diff
@@ -375,6 +375,16 @@ def visit_BinOp(self, node: BinOp) -> Any:
...
 
         if isinstance(node.op, BitOr):
+            # If either branch of the BinOp has been transformed to `None`
+            # then the `ast.generic_visit` will eliminate that branch completely.
+            # If this happens, treat the BinOp as just the other branch.
+            if not hasattr(node, "left") and not hasattr(node, "right"):
+                return None
+            if not hasattr(node, "left"):
+                return node.right
+            if not hasattr(node, "right"):
```
Comment (agronholm) [inline src/typeguard/_transformer.py:385]: This is just for aesthetics, but I prefer separate control blocks to be separated by a blank line.
```suggestion
            elif not hasattr(node, "left"):
                return node.right
            elif not hasattr(node, "right"):
```

**Comment by agronholm** [comment]
Thanks!

File: tests/test_transformer.py (hunk 1 above)
Comment (vthemelis) [reply tests/test_transformer.py:1598]: Added https://github.com/agronholm/typeguard/issues/395
